File: youtube_scraper.py
"""
YouTube Scraper Module
---------------------
What this does (in plain English):
  1. Takes a list of YouTube channel handles from your config file.
  2. For each channel, finds all videos uploaded in the last 7 days.
  3. For each recent video, tries to download the full transcript (captions).
  4. If no transcript is available, it notes that and moves on.
  5. Returns a list of dictionaries, each containing the video title, URL,
     channel name, publish date, and either the transcript or a note that
     it was unavailable.
"""

import datetime
import scrapetube
from youtube_transcript_api import YouTubeTranscriptApi
from youtube_transcript_api._errors import (
    TranscriptsDisabled,
    NoTranscriptFound,
    VideoUnavailable,
)


def get_recent_videos(channel_handle, days=7):
    """
    Get videos from a YouTube channel uploaded within the last `days` days.
    Uses scrapetube which scrapes the channel page — no API key needed.
    """
    cutoff = datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(days=days)
    videos = []

    try:
        # scrapetube returns an iterator of video dicts
        for video in scrapetube.get_channel(channel_url=f"https://www.youtube.com/{channel_handle}"):
            video_id = video["videoId"]
            title = video.get("title", {}).get("runs", [{}])[0].get("text", "Unknown Title")

            # Extract publish time text (e.g. "3 days ago")
            publish_text = ""
            if "publishedTimeText" in video:
                publish_text = video["publishedTimeText"].get("simpleText", "")

            # Filter: only include videos from the last 7 days
            # scrapetube doesn't give exact dates, so we parse relative time
            if not _is_within_days(publish_text, days):
                # Videos are sorted newest-first; once we pass the cutoff, stop
                break

            videos.append({
                "video_id": video_id,
                "title": title,
                "url": f"https://www.youtube.com/watch?v={video_id}",
                "published": publish_text,
            })
    except Exception as e:
        print(f"  [WARNING] Could not fetch videos for {channel_handle}: {e}")

    return videos


def _is_within_days(publish_text, days=7):
    """
    Parse YouTube's relative time strings like '2 days ago', '5 hours ago',
    '1 week ago' to decide if a video is within the window.
    """
    publish_text = publish_text.lower().strip()
    if not publish_text:
        return True  # If we can't tell, include it

    # "Streamed X ago" — strip the "Streamed" prefix
    publish_text = publish_text.replace("streamed ", "")

    if "just now" in publish_text or "second" in publish_text or "minute" in publish_text:
        return True
    if "hour" in publish_text:
        return True
    if "day" in publish_text:
        try:
            num = int(publish_text.split()[0])
            return num <= days
        except (ValueError, IndexError):
            return True
    if "week" in publish_text:
        try:
            num = int(publish_text.split()[0])
            return num <= 1 and days >= 7
        except (ValueError, IndexError):
            return False
    # "month", "year", etc. — too old
    return False


def get_transcript(video_id):
    """
    Attempt to download the transcript for a YouTube video.
    Returns (transcript_text, success_bool).
    """
    try:
        transcript_list = YouTubeTranscriptApi.get_transcript(video_id)
        # Join all text segments into one string
        full_text = " ".join(segment["text"] for segment in transcript_list)
        return full_text, True
    except (TranscriptsDisabled, NoTranscriptFound, VideoUnavailable):
        return None, False
    except Exception as e:
        return None, False


def fetch_youtube_data(channels, days=7):
    """
    Main function called by the orchestrator.
    channels: list of dicts with 'name' and 'channel_handle' keys (from config.yaml)
    Returns: list of video dicts with transcript info.
    """
    all_videos = []

    for ch in channels:
        name = ch["name"]
        handle = ch["channel_handle"]
        print(f"  Scanning YouTube channel: {name} ({handle})...")

        videos = get_recent_videos(handle, days=days)
        print(f"    Found {len(videos)} video(s) from the last {days} days.")

        for v in videos:
            transcript, success = get_transcript(v["video_id"])
            v["channel_name"] = name
            if success:
                v["transcript"] = transcript
                v["transcript_status"] = "available"
                print(f"    [OK] Transcript found: {v['title']}")
            else:
                v["transcript"] = None
                v["transcript_status"] = "unavailable"
                print(f"    [SKIP] No transcript: {v['title']}")

            all_videos.append(v)

    return all_videos

File: youtube_scraper.js
/**
 * YouTube Scraper Module
 * ----------------------
 * What this does (in plain English):
 *   1. Takes a list of YouTube channel handles from your config file.
 *   2. For each channel, scrapes the channel page to find recent video IDs.
 *   3. For each recent video (last 7 days), downloads the transcript (captions).
 *   4. If no transcript is available, it skips that video and notes it.
 *   5. Returns an array of objects with video title, URL, channel, and transcript.
 *
 * No API key needed — this uses public YouTube pages and the free transcript API.
 */

const https = require("https");
const http = require("http");

/**
 * Make an HTTPS/HTTP GET request and return the response body as a string.
 */
function httpGet(url, headers = {}) {
  return new Promise((resolve, reject) => {
    const client = url.startsWith("https") ? https : http;
    const req = client.get(url, { headers }, (res) => {
      // Follow redirects
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return httpGet(res.headers.location, headers).then(resolve).catch(reject);
      }
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => resolve(data));
    });
    req.on("error", reject);
    req.setTimeout(30000, () => {
      req.destroy();
      reject(new Error("Request timed out"));
    });
  });
}

/**
 * Scrape a YouTube channel page to find recent video IDs and titles.
 * We fetch the channel's /videos page HTML and extract data from the
 * embedded JSON (ytInitialData).
 */
async function getRecentVideos(channelHandle, days = 7) {
  const videos = [];
  const url = `https://www.youtube.com/${channelHandle}/videos`;

  try {
    const html = await httpGet(url, {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      "Accept-Language": "en-US,en;q=0.9",
    });

    // Extract ytInitialData JSON from the page
    const match = html.match(/var ytInitialData = ({.*?});<\/script>/s);
    if (!match) {
      console.log(`    [WARNING] Could not parse channel page for ${channelHandle}`);
      return videos;
    }

    const data = JSON.parse(match[1]);

    // Navigate the nested YouTube data structure to find video entries
    const tabs =
      data?.contents?.twoColumnBrowseResultsRenderer?.tabs || [];
    let videoTab = null;
    for (const tab of tabs) {
      if (tab?.tabRenderer?.title === "Videos") {
        videoTab = tab;
        break;
      }
    }

    const items =
      videoTab?.tabRenderer?.content?.richGridRenderer?.contents || [];

    for (const item of items) {
      const renderer = item?.richItemRenderer?.content?.videoRenderer;
      if (!renderer) continue;

      const videoId = renderer.videoId;
      const title =
        renderer.title?.runs?.[0]?.text || "Unknown Title";
      const publishText =
        renderer.publishedTimeText?.simpleText || "";
      const description =
        renderer.descriptionSnippet?.runs?.map((r) => r.text).join("") || "";

      // Filter by recency
      if (!isWithinDays(publishText, days)) {
        break; // Videos are sorted newest-first
      }

      videos.push({
        video_id: videoId,
        title: title,
        url: `https://www.youtube.com/watch?v=${videoId}`,
        published: publishText,
        description: description,
      });
    }
  } catch (err) {
    console.log(
      `    [WARNING] Could not fetch videos for ${channelHandle}: ${err.message}`
    );
  }

  return videos;
}

/**
 * Parse YouTube's relative time strings ("2 days ago", "5 hours ago", etc.)
 */
function isWithinDays(publishText, days = 7) {
  const text = publishText.toLowerCase().replace("streamed ", "").trim();
  if (!text) return true;

  if (
    text.includes("just now") ||
    text.includes("second") ||
    text.includes("minute") ||
    text.includes("hour")
  ) {
    return true;
  }
  if (text.includes("day")) {
    const num = parseInt(text);
    return !isNaN(num) ? num <= days : true;
  }
  if (text.includes("week")) {
    const num = parseInt(text);
    return !isNaN(num) ? num <= 1 && days >= 7 : false;
  }
  return false; // "month", "year", etc.
}

/**
 * Fetch the transcript for a YouTube video using the youtube-transcript package.
 * Returns { text, success }.
 */
async function getTranscript(videoId) {
  try {
    // Use the youtube-transcript npm package
    const { YoutubeTranscript } = require("youtube-transcript");
    const segments = await YoutubeTranscript.fetchTranscript(videoId);
    const fullText = segments.map((s) => s.text).join(" ");
    return { text: fullText, success: true };
  } catch (err) {
    return { text: null, success: false };
  }
}

/**
 * Fetch the full video description from the watch page.
 * The channel listing only has a snippet; this gets the complete text.
 */
async function getVideoDescription(videoId) {
  try {
    const url = `https://www.youtube.com/watch?v=${videoId}`;
    const html = await httpGet(url, {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      "Accept-Language": "en-US,en;q=0.9",
    });

    const match = html.match(/var ytInitialData = ({.*?});<\/script>/s);
    if (!match) return null;

    const data = JSON.parse(match[1]);
    const contents =
      data?.contents?.twoColumnWatchNextResults?.results?.results?.contents || [];

    for (const item of contents) {
      const desc =
        item?.videoSecondaryInfoRenderer?.attributedDescription?.content;
      if (desc) return desc;

      // Alternative location in some page layouts
      const metaDesc =
        item?.videoSecondaryInfoRenderer?.description?.runs;
      if (metaDesc) return metaDesc.map((r) => r.text).join("");
    }
  } catch {
    // Silently fail — description is a best-effort fallback
  }
  return null;
}

/**
 * Main function called by the orchestrator.
 * channels: array of { name, channel_handle } from config.json
 * Returns: array of video objects with transcript info.
 */
async function fetchYoutubeData(channels, days = 7) {
  const allVideos = [];

  for (const ch of channels) {
    console.log(`  Scanning YouTube channel: ${ch.name} (${ch.channel_handle})...`);
    const videos = await getRecentVideos(ch.channel_handle, days);
    console.log(`    Found ${videos.length} video(s) from the last ${days} days.`);

    for (const v of videos) {
      const { text, success } = await getTranscript(v.video_id);
      v.channel_name = ch.name;

      if (success) {
        v.transcript = text;
        v.transcript_status = "available";
        console.log(`    [OK] Transcript found: ${v.title}`);
      } else {
        // Try to get the full description from the video watch page as fallback
        const desc = await getVideoDescription(v.video_id);
        v.transcript = null;
        v.description = desc || v.description || null;
        v.transcript_status = "unavailable";
        if (desc) {
          console.log(`    [DESC] No transcript, using video description: ${v.title}`);
        } else {
          console.log(`    [SKIP] No transcript or description: ${v.title}`);
        }
      }

      allVideos.push(v);
    }
  }

  return allVideos;
}

module.exports = { fetchYoutubeData };
